First early version

hsablonniere opened the pull request:
This is a proposition for a `0.1.0` version. It addresses #3, #4, #5 and #6.

## Content of the PR

* An MIT license
* A new README
  * Do we need the details about how notes are authored? If we do, we'll also have to think about deploying the bespoke plugin and explain that this tool only works with bespoke (for now).
* A globally installable command line tool `prz` (instead of the npm/yarn script)
* Some file cleanups (I removed a lot of files that are WIP, there are still in a branch, we'll use them later)

NOTE: Some renaming will be needed between components and the names we use in the doc (presentation console & projector viewport).

## TODO (after merge)

* [x] Rename the project to `ensuite/ensuite-present`
* [x] Close #4, #5 and #6
* [ ] Tag/version the release
* [x] Publish on npm

hsablonniere committed:
first commit with main design concept in place

* pages
* components
* remote configs
* channels and bus

hsablonniere committed:
change notes indent to exactly 2 characters

hsablonniere committed:
add architecture diagram

hsablonniere committed:
start stopwatch and timer at the same time (sfc30)

hsablonniere committed:
add ratio controls (keyboard)

hsablonniere committed:
add special move (logitech remote)

hsablonniere committed:
improve footer readability

hsablonniere committed:
display black screen when viewer needs to display something before first step

hsablonniere committed:
improve notes readability

hsablonniere committed:
add missing remotes

hsablonniere committed:
fix notes scroll position on resize

hsablonniere committed:
add notes, questions and thoughts

hsablonniere committed:
cleanup files that aren't necessary (for now) for the release

hsablonniere committed:
replace npm script with proper globally installable CLI tool

hsablonniere committed:
add the MIT license ;-)

hsablonniere committed:
enrich the README with more instructions and context

hsablonniere committed:
add restrictions on node/npm versions

mojavelinux commented on README.adoc:
```diff
@@ -1,3 +1,45 @@
+= ensuite-present
 
-This project is a Work In Progress
+_ensuite-present_ is part of the *ensuite* toolbelt dedicated to trainers and speakers.
```
Let's call it a toolkit.

mojavelinux commented on README.adoc:
```diff
@@ -1,3 +1,45 @@
+
+== Overview
+
+_ensuite-present_ is a local Web app that allows an instructor to "open" and control a slide-deck.
```
Let's lowercase Web here. There are different opinions, but I think it makes it more readable.

hsablonniere replied:
I had to try :trollface:

mojavelinux commented on README.adoc:
```diff
@@ -1,3 +1,45 @@
+
+To run _ensuite-present_ on your machine, you just need to run this command:
+
+ $ prz
```
I know I keep going back and forth on this, but I wonder if we should name it prez, es-prez, or es-present instead of prz. Seeing it alone here kind of gives me a feeling of wanting to change the name. We don't have to get it right for this release, just something to think about before 1.0.0.

hsablonniere replied:
I would defer that to another release.

mojavelinux replied:
I'm happy to defer. My brain is really getting attached to "enpres" though. But no need to rush it. We have time to let it soak.

mojavelinux commented on bin/ensuite-present.js:
```diff
@@ -0,0 +1,23 @@
+const http = require('http')
+const nodeStatic = require('node-static')
+
+const PORT = 4242
```
4242 is the port used by Awestruct. I think we should use 4320 instead (since we're proposing 4321 for the bespoke generator).

hsablonniere replied:
I still think bespoke should propose a pseudo-random port but that's another disussion :smile:

hsablonniere replied:
We'll make `4320` configurable in later releases.

mojavelinux commented on bin/ensuite-present.js:
```diff
@@ -0,0 +1,23 @@
+const fileServer = new nodeStatic.Server()
+
+http
+  .createServer((request, response) => {
```
When installed, this seems to be serving files from the bin directory instead of from ensuite-present.

mojavelinux replied:
We need something like:

```js
const fileServer = new nodeStatic.Server(path.dirname(path.dirname(require.resolve('ensuite-present'))))
```

hsablonniere replied:
I implemented it like this:

```js
const webRoot = path.resolve(__dirname, '..')
const fileServer = new nodeStatic.Server(webRoot)
```

mojavelinux reviewed (changes requested):
Can we dim the next slide? Right now it's hard to distinguish the state of the two slides.

I think we should put a placeholder word when a slide has no notes, which we can do with CSS. Otherwise, I find it confusing what the cursor is pointing to. Something like "(no notes)" or "-" would work.

```css
.notes-line:empty::after {
  content: '(no notes)';
  font-style: italic;
  opacity: 0.33;
}
```

I don't think we should highlight the notes on hover, or at least not so prominently. Perhaps place a border or outline to indicate the boundaries, but the dark background color is confusing.

I think the text in the notes viewer needs to be a little tighter. Here are the settings I propose:

* font-size: 28px
* padding: 0.5em 1em;
* hanging indent: 1.25ex

I think we should hide the scrollbar on Chrome / IE using browser-specific styles. See https://github.com/opendevise/presentation-bespoke-starter/blob/master/src/styles/bespoke-modes.styl#L2-L3

I'd like to have a button to hide the timer bar, but that can be part of a follow-up release.

mojavelinux commented:
> Do we need the details about how notes are authored?

Not at this time. Perhaps just make a brief reference to https://github.com/asciidoctor/asciidoctor-bespoke#speaker-notes and that more info is coming.

mojavelinux commented:
I think we should simplify the page URLs. Right now it's:

* pages/slide-deck-viewer/slide-deck-viewer-page.html
* pages/teleprompter-console/teleprompter-console-page.html

I think this could be simplified to:

* pages/viewer/index.html
* pages/console/index.html

I also think the query string parameter should be shorten from `slide-deck-url` to `load` or `deck` or `url`.

mojavelinux commented:
Another idea for the tool names is to prefix them with "en" and replace "suite" with the name of the tool. For example, "enpres". Roughly translated, that means something related to a business venture in some languages, which is somewhat interesting.

I may be overthinking this. Names tend to create that problem.

hsablonniere committed:
make empty notes visible with a half transparent "(no notes)" line

I had to improve a bit the trimming/filtering of the notes.

hsablonniere committed:
change visual indicator of hovered note to a left border

hsablonniere committed:
compact notes a bit

hsablonniere committed:
typos and quick fix in readme

hsablonniere committed:
change port from 4242 to 4320

hsablonniere committed:
fix path from where static web files are served

hsablonniere commented:
> Can we dim the next slide? Right now it's hard to distinguish the state of the two slides.

I would prefer not to implement "distinguishability" with dimming. It's one of the reasons I chose to make the next slide as big as the current one compared to other solutions (Reveal, PowerPoint, Keynote...): **it needs to be as readable as the current slide**.

At some point I had a line between the two slides, maybe we could try that again? I think positions should be enough to get use to the fact that upper one is current slide and lower one is next slide.

hsablonniere commented:
> I think we should put a placeholder word when a slide has no notes...

I already tried it and was not convinced but I implemented your proposition in a commit anyway. :smile:

hsablonniere commented:
> I don't think we should highlight the notes on hover...

I think we need to keep hover indication, it helps discovering clickability and if I hover a note that is not the current one, I can see clearly how long it is.

I implemented something with a border, let me know what you think...

![capture d ecran_2017-02-14_11-50-07](https://cloud.githubusercontent.com/assets/236342/22927181/3755ee76-f2b0-11e6-9be3-5adadfea8e05.png)

hsablonniere commented:
> I think the text in the notes viewer needs to be a little tighter. Here are the settings I propose:

* OK for `font-size: 28px`.
* OK for `padding: 0.5em 1em`, assuming you're talking about `.notes-block`.
* What about hanging indent? `ex` are relative to height, I wanted to characters indent thus the use of `2ch`. Would you prefer 1ch. I tried it's not that great.

hsablonniere commented:
> I think we should hide the scrollbar on Chrome / IE using browser-specific styles.

I really want to keep it. Removing that kind of scrollbars usually comes with usability problems. I used that a few times when looking through notes to find a slide. I also used it as a visual indicator of progress during the talk.

hsablonniere commented:
> I'd like to have a button to hide the timer bar, but that can be part of a follow-up release.

Let's discuss that in an issue :smile:

hsablonniere commented:
> Not at this time. Perhaps just make a brief reference to https://github.com/asciidoctor/asciidoctor-bespoke#speaker-notes and that more info is coming.

The think is that I really want it to work with any kind of notes system.
Whether is asciidoctor-bespoke, bespoke with jade, reveal...

hsablonniere commented:
> I think we should simplify the page URLs.

I also think page URLs are too long, but at the same time, I really want to enforce this rule: "all file should have specific names". It really helps productivity in your IDE when you're looking of a file.

We can achieve what you want with some kind of router but if we agree on that I would also defer that to another release and PR.

hsablonniere commented:
> I also think the query string parameter should be shorten from slide-deck-url to load or deck or url.

I like prefer things over abbreviations. In the end, this parameter will disappear and/or be automatically setup with a prompt or a deck grid view to select the one.

hsablonniere commented:
> Another idea for the tool names is to prefix them with "en" and replace "suite" with the name of the tool. For example, "enpres". Roughly translated, that means something related to a business venture in some languages, which is somewhat interesting.

Let's move discussion about naming out from this early release.

mojavelinux commented:
> At some point I had a line between the two slides, maybe we could try that again?

Let's do that for now. I think the problem is with white slides, it just looks like one slide. So, in that case, it's actually harder to read.

I think just a horizontal line or gap would suffice.

Another idea, which we don't have to implement or try it now, is to slightly reduce the size of the slide so it not only has a border around it, but has an appearance of coming into view. Another idea would be a slight perspective tilt so it looks like it's on a carousel coming up from the floor. I'm not talking about a major rotation. Just enough to give you a subconscious hint. Just an idea.

mojavelinux commented:
> I implemented something with a border (for the note hover), let me know what you think...

That's exactly what we want. Perfect.

mojavelinux commented:
> What about hanging indent? ex are relative to height, I wanted to characters indent thus the use of 2ch. Would you prefer 1ch. I tried it's not that great.

I think due to font differences we are seeing different behavior. It's not a huge deal. With my font (which I think is Liberation Mono from Fedora 21), 1.25 was lining up at right about 2 characters. I agree that 1ch is too small.

(I'm pretty sure ex and ch are the same, at least from what I'm observing...so I didn't mean to propose a unit change).

mojavelinux commented:
> The think is that I really want it to work with any kind of notes system.
Whether is asciidoctor-bespoke, bespoke with jade, reveal...

I totally agree. And we do need to document it formally. I was just trying to save you some time by saying that we explain how the notes get populated in the Asciidoctor Bespoke converter and that could at least give users an idea of what we're doing until we have time to write about it. But if you'd like to throw a few sentences together in this README, go for it.

mojavelinux commented:
> Let's move discussion about naming out from this early release.

Agreed.

mojavelinux commented:
Scrollbars stay in.

mojavelinux commented:
I'm okay with keeping the query string parameter as it is. But I still think the page names are too specific and actually making it hard to document. Something a lot shorter makes it easier to explain because the terminology doesn't get in the way. I could go a bit more specific though. How about:

* presenter-console (or speaker console)
* observer-view (or projector-view or viewfinder)

(viewfinder was my last idea, but has a very interesting connotation as a read-only point of view).

hsablonniere committed:
add borders between slides and notes

hsablonniere committed:
make note hover indicator more visible

after adding black borders between notes and slides, it was less visible

hsablonniere commented:
> Let's do that for now. I think the problem is with white slides, it just looks like one slide. So, in that case, it's actually harder to read. I think just a horizontal line or gap would suffice.

I've also implemented a vertial line with the notes viewer. It looks more consistent.

![capture d ecran_2017-02-15_15-12-20](https://cloud.githubusercontent.com/assets/236342/22988326/fba6fbce-f3b1-11e6-9940-fa8a5b3f746a.png)

hsablonniere commented:
> Another idea, which we don't have to implement or try it now, is to slightly reduce the size of the slide so it not only has a border around it, but has an appearance of coming into view. Another idea would be a slight perspective tilt so it looks like it's on a carousel coming up from the floor. I'm not talking about a major rotation. Just enough to give you a subconscious hint. Just an idea.

I'll let you create an issue for this to discuss about it. I'm not sure I see what you mean.

hsablonniere commented:
> I think due to font differences we are seeing different behavior. It's not a huge deal. With my font (which I think is Liberation Mono from Fedora 21), 1.25 was lining up at right about 2 characters. I agree that 1ch is too small.

Could you provide a screenshot?

hsablonniere commented:
> (I'm pretty sure ex and ch are the same, at least from what I'm observing...so I didn't mean to propose a unit change).

They're different :smile:

* `ex` unit is based on the x-height of the current font. ([details](https://css-tricks.com/the-lengths-of-css/#article-header-id-10))
* `ch` unit is based on the width of the `0` character. ([details](https://css-tricks.com/the-lengths-of-css/#article-header-id-11))

hsablonniere commented:
> I totally agree. And we do need to document it formally. I was just trying to save you some time by saying that we explain how the notes get populated in the Asciidoctor Bespoke converter and that could at least give users an idea of what we're doing until we have time to write about it. But if you'd like to throw a few sentences together in this README, go for it.

I added a few lines about it but it relies on the protocol and the bespoke-plugin which both are undocumented, alpha  and not published.

hsablonniere committed:
mentions a few things about notes

hsablonniere committed:
simplify pages naming and URLs while keeping unique filenames and type suffix

hsablonniere commented:
> I still think the page names are too specific and actually making it hard to document.

Those names were specific because I want to be explicit. I see more than one console in the future so I added teleprompter. The suffix `-page` is here to organize stuffs properly. In order to move forward, I renamed them to:

* `/pages/console/console-page.html`
* `/pages/viewer/viewer-page.html`

I added a bit of "routing" to have them accessible as URLs at:

* `http://localhost:4320/pages/viewer/`
* `http://localhost:4320/pages/console/?slide-deck-url=http://localhost:4321/my-awesome-slide-deck.html`

hsablonniere commented:
About the naming we'll move to #3.

mojavelinux commented:
Okay, I see that ex and ch are different, but can act the same depending on the font (if x height matches 0 width, which is true in some cases). I'm 100% in support for ch for this usage.

mojavelinux commented:
I like the new page names and the router. Perfect.

> I see more than one console in the future so I added teleprompter.

Once we get to that point, we can certainly rename pages. As you said, the user goes to where we tell them to go, so it's not a huge deal to rename pages in the future. And when the use case comes up, it should be clear what names we need to use.

mojavelinux committed:
update package description

mojavelinux committed:
reduce border around slides

mojavelinux committed:
set color instead of opacity for empty notes; use VIM-style blank line

mojavelinux committed:
don't add hover border for current notes; make border lighter

mojavelinux committed:
remove space in shebang line

mojavelinux committed:
set web root correct when running from local installation

mojavelinux committed:
fix URL router

mojavelinux committed:
print URLs to viewer and console on start

mojavelinux committed:
update description of ensuite

mojavelinux committed:
update package description

mojavelinux committed:
apply revisions to README

mojavelinux committed:
change bin name to enprez

mojavelinux commented:
I made a small update so that the left border on the notes does not show for the current notes since they are already highlighted in yellow.

I also changed to the more subtle VIM-style no content. (Just a tilde).

mojavelinux committed:
fix titles of windows

hsablonniere commented:
I rewrote history to remove a forgotten comment in the VIM style commit.

hsablonniere commented:
I think we'll have to improve implementation on:

* The "router" obviously
* The `__dirname` try/catch

mojavelinux commented:
I agree the calculation of the web root is not pretty. However, it currently allows the script to work in dev mode (from inside the repository) or installed mode (installed as a bin script with a loadable module). If there's a way to do this correctly in the node/npm ecosystem, I'd love to know. However, I could not find a way other than what I came up with. The catch will never be hit in installed mode, so for me that's pretty safe.

Btw, it's typical to use this trick in the Ruby ecosystem, so for me it's natural. One other approach is to try to detect if the script is relative to another known file. If so, we know what mode we're in.

mojavelinux commented:
My statement about `ch` units above was incorrect. I was a victim of my own optical illusion. As it turns out, the characters on different lines are, in fact, lining up correctly.

mojavelinux commented:
I looks like we're still missing a tag in the repository. Can you add it?

hsablonniere commented:
😱 Oops, I pushed it to my forked instead of this remote.

https://github.com/ensuite/ensuite-present/releases/tag/v0.1.0

hsablonniere commented:
> I agree the calculation of the web root is not pretty. However, it currently allows the script to work in dev mode (from inside the repository) or installed mode (installed as a bin script with a loadable module). If there's a way to do this correctly in the node/npm ecosystem, I'd love to know. However, I could not find a way other than what I came up with. The catch will never be hit in installed mode, so for me that's pretty safe.

We'll hack a look later. As we said many times (or will say): "make it until your improve it!".
